---
title: "Expression data: TPM median to soybean and common bean"
author: "Turquetti-Moraes, Dayana Kelly"
date: '2024'
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Loading libraries 

```{r}
library(BioNERO)
library(tidyverse)
library(pheatmap)
library(RColorBrewer)
library(readr)
library(reshape2)

```

## Define functions

```{r}

#calc median
median_calc <- function (x) {
  apply(x, 1, median, na.rm=TRUE)
}

# Add median calculated to a column
add_median_column <- function(df, name) {
  df <- mutate(df, median = median_calc(df))
  df <- select(df, median)
  return(df)
}

#Calc TPM values of raw expression data of common bean
calc.tpm <- function(data) {
  
  if(!("length" %in% names(data))){
    stop("Error: column name 'length' not found in names(data)")
  }
  feature_length <- data[,"length",drop=FALSE]
  counts <- data[,!(names(data)=="length"),drop=FALSE]
  
  ##Calculate the RPK value
  RPK <- matrix(0, nrow=dim(counts)[1], ncol=dim(counts)[2])
  
  for(row in 1:dim(counts)[1]){
    for(col in 1:dim(counts)[2]){
      RPK[row,col] <- counts[row,col]/feature_length$length[row]
    }
  }
  
  ##Calculate the sums of each column and divide by 1000000
  scale_factor <- colSums(RPK)/1000000
  
  ##Now divide all values in each column by the scaling factor
  TPM <- t(t(RPK)/scale_factor)
  colnames(TPM) <- names(counts)
  row.names(TPM) <- row.names(counts)
  return(as.data.frame(TPM))
}

#Plot heatmap
heatmap_plot <- function(x) {
  
  my_color_palette <- RColorBrewer::brewer.pal(n=4, name= "Reds")
  limite_cores <- c(0, 7.99999, 20, 30, 40)

  hm <- pheatmap(as.matrix(x),
    color= my_color_palette,
    border_color = "white",
    cluster_cols = FALSE,
    cluster_rows = FALSE,
    show_rownames = TRUE,
    breaks = limite_cores,
    cellwidth = 10,
    fontsize = 6,
    height = 20,
    cutree_cols = 1) 
  
  return(hm)
}

```

## Selected samples from Soybean Expression Atlas (sea)
The Soybean Expression Atlas v2 is available at https://soyatlas.venanciogroup.uenf.br/

```{r}

#Selected samples from sea: sea_selected_tpm (large file not available)

#Metadata: metadata_sea (not available)

#Results are in df_medians (available in .Rdata)

```

## Join Parts and BioSamples

```{r}
# Group the data by plant part and BioSample
groups <- split(metadata_sea$BioSample, metadata_sea$Part)

# Create objects in the R environment for each group
for (group_name in names(groups)) {
  assign(group_name, groups[[group_name]])
}

# Print the BioSamples for each Part group
for (group_name in names(groups)) {
  cat(group_name, "(", paste(groups[[group_name]], collapse = ", "), ")\n")
}

```

## Select expression data for each part

```{r}

# Make a dataframe for each group of plant part 
for (group_name in names(groups)) {
  selected_columns <- c(groups[[group_name]])
  
  # Dataframe with selected samples
  assign(paste0(group_name, "_exp"), sea_selected_tpm[, selected_columns])
}

# Visualize a specific dataframe of a plant part expression
print(cotyledon_exp)

```

## Calc expression median

```{r}

# List of the plant parts expression data
part_exp <- list(seedcoat_exp, embryo_exp, seed_exp, cotyledon_exp, endosperm_exp, leaf_exp, epicotyl_exp, hypocotyl_exp, shoot_exp, suspensor_exp, flower_exp, pod_exp, nodule_exp, root_exp) 

# Using the function add_median_column to calc the median of expression data from each plant part
medians <- map(part_exp, add_median_column)

# Convert a list to a dataframe
df_medians <- bind_cols(medians)
colnames(df_medians) <- c("SeedCoat", "Embryo", "Seed", "Cotyledon", "Endosperm", "Leaf", "Epicotyl", "Hypocotyl", "Shoot", "Suspensor", "Flower", "Pod", "Nodule", "Root")

```

## Seed expression to soybean and classification of expression intensity

```{r soybean seed expression}

gma_seed_exp <- df_medians %>% select(SeedCoat, Embryo, Seed, Cotyledon, Endosperm)

# Remove those genes without seed expression
gma_seed_exp$Sum <- rowSums(gma_seed_exp)

# All values in a line are <1?
remove_lines <- apply(gma_seed_exp[, -ncol(gma_seed_exp)], 1, function(x) all(x < 1))

# Remove all lines with expression <1
gma_seed_exp <- gma_seed_exp[!remove_lines, ]
gma_seed_exp <- gma_seed_exp[, -ncol(gma_seed_exp)]

```

## Classify expression intensity to genes expressed in gma seeds

```{r}

#Calc max mediam value for each gene
gma_seed_exp$Max_exp <- apply(gma_seed_exp, 1, max)

# Classify seed expression intensity
gma_seed_exp <- gma_seed_exp %>%
  mutate(Intensity_exp = case_when(
    Max_exp <= 5 ~ "low",
    Max_exp > 5 & Max_exp < 10 ~ "medium",
    Max_exp >= 10 ~ "high",
    TRUE ~ NA_character_
  ))

```

## Gma seed expression: at least 8 TPM

```{r}

gma_seed_exp_8tpm <- gma_seed_exp %>%
  filter(Max_exp >= 8)
  
```

## Gma seed expression in candidate homologous families

```{r}

#Expressed genes in candidate homologous families
gma_seed_exp_colname <- rownames_to_column(gma_seed_exp, var="GeneID") 
gma_seed_exp_homofam <- inner_join(gma_seed_exp_colname, oil_homofam, by="GeneID") 

```

## Gma seed expression in candidate homologous families: at least 8 TPM

```{r}

#Expressed genes in candidate homologous families, at least 8 TPM
gma_seed_exp_8tpm_colname <- rownames_to_column(gma_seed_exp_8tpm, var="GeneID")
gma_seed_exp_8tpm_homofam <- inner_join(gma_seed_exp_8tpm_colname, oil_homofam, by="GeneID") 

```

## Soybean expression of TAG synthesis pathway genes

```{r}

#gma TAG synthesis
df_medians_colname <- rownames_to_column(df_medians, var="GeneID")
gma_tag_synthesis_exp <- inner_join(gma_tag_synthesis, df_medians_colname, by="GeneID") 
gma_tag_synthesis_exp <- column_to_rownames(gma_tag_synthesis_exp, var="GeneID")

#transform data to plot: TPM > 40 = 40
gma_tag_synthesis_exp_40 <- gma_tag_synthesis_exp %>%
  mutate_all(~ifelse(. > 40, 40, .))

#Plot heatmap
heatmap_plot(gma_tag_synthesis_exp_40)

```

## Soybean expression of TAG synthesis pathway genes: at least 8 TPM

```{r}

gma_tag_synthesis_exp_40_colname <- rownames_to_column(gma_tag_synthesis_exp_40, var="GeneID")
gma_genes_8tpm <- gma_seed_exp_8tpm_colname %>% select(GeneID)
gma_tag_synthesis_exp_40_8tpm <- inner_join(gma_genes_8tpm, gma_tag_synthesis_exp_40_colname, by="GeneID")

#Plot heatmap
gma_tag_synthesis_exp_40_8tpm <- column_to_rownames(gma_tag_synthesis_exp_40_8tpm, var="GeneID")

heatmap_plot(gma_tag_synthesis_exp_40_8tpm)

```

## Soybean expression of TAG degradation pathway genes

```{r}

#gma TAG degradation
gma_tag_degradation_exp <- inner_join(gma_tag_degradation, df_medians_colname, by="GeneID") 
gma_tag_degra(~ifelse(. > 40, 40, .))

#Plot heatmap
heatmap_plot(gma_tag_degradation_exp_40)

```

## Soybean expression of TAG degradation pathway genes: at least 8 TPM

```{r}

gma_tag_degradation_exp_40_colname <- rownames_to_column(gma_tag_degradation_exp_40, var="GeneID")
gma_genes_8tpm <- gma_seed_exp_8tpm_colname %>% select(GeneID)
gma_tag_degradation_exp_40_8tpm <- inner_join(gma_genes_8tpm, gma_tag_degradation_exp_40_colname, by="GeneID")

#Plot heatmap
gma_tag_degradation_exp_40_8tpm <- column_to_rownames(gma_tag_degradation_exp_40_8tpm, var="GeneID")

heatmap_plot(gma_tag_degradation_exp_40_8tpm)

```

## Soybean seed specific expression: at least 8 tpm

```{r}
#gma_seed_specific_8tpm is available in .Rdata 

#transform data to plot: TPM > 40 = 40
gma_seed_specific_8tpm_40 <- gma_seed_specific_8tpm %>%
  mutate_all(~ifelse(. > 40, 40, .))

#Plot heatmap
heatmap_plot(gma_seed_specific_8tpm_40)
```

## Common bean expression data

The common bean expression data used in this paper is from: 
O’Rourke, Jamie A., Luis P. Iniguez, Fengli Fu, Bruna Bucciarelli, Susan S. Miller, Scott A. Jackson, Philip E. McClean, et al. 2014. “An RNA-Seq Based Gene Expression Atlas of the Common Bean.” BMC Genomics 15 (1): 866. https://doi.org/10.1186/1471-2164-15-866.

From pvu raw expression data to TPM normalization

```{r}
# PVU raw expression counts and gene lengths are available in .Rdata: Pvu_raw_expression_count_gene_length

# Calc TPM of common bean raw expression data

TPM_pvu <- calc.tpm(Pvu_raw_expression_count_gene_length)
TPM_pvu_colname <- rownames_to_column(TPM_pvu, var="GeneID")

```

## Seed expression to common bean and classification of expression intensity

```{r common bean seed expression}

pvu_seed_exp <- TPM_pvu %>% select(`PvSH-1`, `PvS1-1`, `PvS2-1`)

# Remove those genes without seed expression
pvu_seed_exp$Sum <- rowSums(pvu_seed_exp)

# All values in a line are <1?
remove_lines_pvu <- apply(pvu_seed_exp[, -ncol(pvu_seed_exp)], 1, function(x) all(x < 1))

# Remove all lines with expression <1
pvu_seed_exp <- pvu_seed_exp[!remove_lines_pvu, ]
pvu_seed_exp <- pvu_seed_exp[, -ncol(pvu_seed_exp)]

```

## Classify expression intensity to genes expressed in pvu seeds

```{r}
#Calc max mediam value for each gene

pvu_seed_exp$Max_exp <- apply(pvu_seed_exp, 1, max)

# Classify seed expression intensity

pvu_seed_exp <- pvu_seed_exp %>%
  mutate(Intensity_exp = case_when(
    Max_exp <= 5 ~ "low",
    Max_exp > 5 & Max_exp < 10 ~ "medium",
    Max_exp >= 10 ~ "high",
    TRUE ~ NA_character_  
  ))

```

## Pvu seed expression: at least 8 TPM

```{r}

pvu_seed_exp_8tpm <- pvu_seed_exp %>%
  filter(Max_exp >= 8)

```


## Pvu seed expression in candidate homologous families

```{r}
#Expressed genes in candidate homologous families
pvu_seed_exp_colname <- rownames_to_column(pvu_seed_exp, var="GeneID") 
pvu_seed_exp_homofam <- inner_join(pvu_seed_exp_colname, oil_homofam, by="GeneID") 

#Expressed genes in candidate homologous families, at least 8 TPM
pvu_seed_exp_8tpm_colname <- rownames_to_column(pvu_seed_exp_8tpm, var="GeneID")
pvu_seed_exp_8tpm_homofam <- inner_join(pvu_seed_exp_8tpm_colname, oil_homofam, by="GeneID")  

```

## Common bean expression TAG synthesis pathway genes

```{r}
#pvu_tag_synthesis is available in .Rdata

#pvu TAG synthesis
pvu_tag_synthesis_exp <- inner_join(pvu_tag_synthesis, TPM_pvu_colname, by="GeneID") 
pvu_tag_synthesis_exp <- column_to_rownames(pvu_tag_synthesis_exp, var="GeneID")

#transform data to plot: TPM > 40 = 40
pvu_tag_synthesis_exp_40 <- pvu_tag_synthesis_exp %>%
  mutate_all(~ifelse(. > 40, 40, .))

#Plot heatmap
heatmap_plot(pvu_tag_synthesis_exp_40)

```

## Common bean expression of TAG synthesis pathway genes: at least 8 TPM

```{r}

pvu_tag_synthesis_exp_40_colname <- rownames_to_column(pvu_tag_synthesis_exp_40, var="GeneID")
pvu_genes_8tpm <- pvu_seed_exp_8tpm_colname %>% select(GeneID)
pvu_tag_synthesis_exp_40_8tpm <- inner_join(pvu_genes_8tpm, pvu_tag_synthesis_exp_40_colname, by="GeneID")

#Plot heatmap
pvu_tag_synthesis_exp_40_8tpm <- column_to_rownames(pvu_tag_synthesis_exp_40_8tpm, var="GeneID")

heatmap_plot(pvu_tag_synthesis_exp_40_8tpm)
```

## Common bean expression of TAG degradation pathway genes

```{r}
#pvu TAG degradation: pvu_tag_degradation data is available in .Rdata

#transform data to plot: TPM > 40 = 40
pvu_tag_degradation_exp_40 <- pvu_tag_degradation_exp %>%
  mutate_all(~ifelse(. > 40, 40, .))

#Plot heatmap
heatmap_plot(pvu_tag_degradation_exp_40)

```

## Common bean expression of TAG degradation pathway genes: at least 8 TPM

```{r}

pvu_tag_degradation_exp_40_colname <- rownames_to_column(pvu_tag_degradation_exp_40, var="GeneID")
pvu_genes_8tpm <- pvu_seed_exp_8tpm_colname %>% select(GeneID)
pvu_tag_degradation_exp_40_8tpm <- inner_join(pvu_genes_8tpm, pvu_tag_degradation_exp_40_colname, by="GeneID")

#Plot heatmap
pvu_tag_degradation_exp_40_8tpm <- column_to_rownames(pvu_tag_degradation_exp_40_8tpm, var="GeneID")

heatmap_plot(pvu_tag_degradation_exp_40_8tpm)

```

