---
title: "Coexpression of genes with at least 8 TPM from gain oil homologous families"
author: "Turquetti-Moraes,Dayana Kelly"
date: "2024"
output: html_document
---
```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Gene coexpression network from ATLAS v2 - Selected samples = 605

```{r}

library(BioNERO)
library(tidyverse)
library(SummarizedExperiment)

```

#Data from Soybean Expression Atlas (sea)
The Soybean Expression Atlas v2 is available at https://soyatlas.venanciogroup.uenf.br/

Metadata of selected samples from sea

```{r}

metadata_new 

metadata_summary 

```


#Selected samples

```{r}

samples_new

```

#Select genes of interest

```{r}

TPM_gene_samples

```

#Only oil candidate genes

```{r}

oil_genes_gain_8tpm 

TPM_oil_genes_gain_8tpm_samples <- inner_join(TPM_gene_samples, oil_genes_gain_8tpm, by="GeneID")
TPM_oil_genes_gain_8tpm_samples <- column_to_rownames(TPM_oil_genes_gain_8tpm_samples, var="GeneID")
```

#Preprocess

```{r}

exp_tpm_preprocess_all_genes <- exp_preprocess(TPM_oil_genes_gain_8tpm_samples)
#removed samples: 35

```

#Exploratory data analysis

```{r plot_heatmap, fig.width=6, fig.height=4}

a <- plot_heatmap(exp_tpm_preprocess_all_genes, type = "samplecor")

b <- plot_heatmap(exp_tpm_preprocess_all_genes, type = "expr",use_raster=TRUE)

```


# Principal component analysis (PCA)

```{r pcaplot, fig.small=TRUE}

pre_analysis<- plot_PCA(exp_tpm_preprocess_all_genes, metadata=metadata_summary, PCs = "2x3")

pre_analysis

```

#Gene coexpression network inference

```{r sft_fit}

sft_all <- SFT_fit(exp_tpm_preprocess_all_genes, net_type = "signed hybrid", cor_method = "pearson")
sft_all$power
power_all <- sft_all$power

```

##Plot sft

```{r plot_sft, fig.width=6}

sft_all$plot

```

#Use the power calculated by `SFT_fit()` to infer the GCN. 

```{r exp2net}

net_all <- exp2gcn(exp_tpm_preprocess_all_genes, 
               SFTpower=sft_all$power)

names(net_all)

```

##Explore the coexpression results

```{r dendro}

# Dendro and colors
dendro<-plot_dendro_and_colors(net_all)

```

```{r eigengene_net, fig.height=6}

# Eigengene networks
eigengene<-plot_eigengene_network(net_all)

```

```{r genes_per_module, fig.wide=TRUE}

genes_per_module<-plot_ngenes_per_module(net_all)
genes_per_module

```

#Module-trait associations

```{r moddtraitcor, fig.width=15, fig.height=15}

MEtrait<-module_trait_cor(exp_tpm_preprocess_all_genes, metadata=metadata_summary, MEs=net_all$MEs, cor_method = "pearson", cex.lab.x = 0.7, cex.lab.y = 0.8, cex.text = 1.2, transpose=TRUE)

MEtrait_sig_005 <- filter(MEtrait, pvalue < 0.05)

```

##Hub gene identification

```{r}

hubs <- get_hubs_gcn(exp_tpm_preprocess_all_genes, net_all)
head(hubs)

```

## References

Almeida-Silva, F., Venancio, T.M. BioNERO: an all-in-one R/Bioconductor package for comprehensive and easy biological network reconstruction. Funct Integr Genomics 22, 131-136 (2022).
https://doi.org/10.1007/s10142-021-00821-9

Wickham H, Averick M, Bryan J, Chang W, McGowan LD, François R, Grolemund G, Hayes A, Henry L, Hester J, Kuhn M, Pedersen TL, Miller E, Bache SM, Müller K, Ooms J, Robinson D, Seidel DP, Spinu V, Takahashi K, Vaughan D, Wilke C, Woo K, Yutani H (2019). “Welcome to the tidyverse.” Journal of Open Source Software, 4(43), 1686. doi:10.21105/joss.01686.

Morgan M, Obenchain V, Hester J, Pagès H (2023). SummarizedExperiment: SummarizedExperiment container. doi:10.18129/B9.bioc.SummarizedExperiment, R package version 1.32.0, https://bioconductor.org/packages/SummarizedExperiment.
